crates/circuit/src/packed_instruction.rs

```diff
@@ -39,132 +42,375 @@ enum PackedOperationType {
-    const POINTER_MASK: usize = usize::MAX ^ Self::DISCRIMINANT_MASK;
-
-    /// Extract the discriminant of the operation.
     #[inline]
     fn discriminant(&self) -> PackedOperationType {
-        ::bytemuck::checked::cast((self.0 & Self::DISCRIMINANT_MASK) as u8)
+        // SAFETY: we read (just!) the discriminant from any of the union's members,
+        // since we guarantee it is found in the same place for all bitfields.
+        unsafe { self.gate.discriminant() }
+    }
+}
+
+impl From<StandardGateBits> for BitField {
+    fn from(gate: StandardGateBits) -> Self {
+        Self { gate }
+    }
+}
+
+impl From<StandardInstructionBits> for BitField {
+    fn from(instruction: StandardInstructionBits) -> Self {
+        Self { instruction }
+    }
+}
+
+impl From<PointerBits> for BitField {
+    fn from(pointer: PointerBits) -> Self {
+        Self { pointer }
+    }
+}
+
+// Used by bitfield-struct to convert into and out of the bitfield.
+impl StandardGate {
+    const fn into_bits(self) -> u8 {
+        self as _
+    }
+    const fn from_bits(value: u8) -> Self {
+        if value as usize >= STANDARD_GATE_SIZE {
+            panic!("unexpected standard gate type!")
+        }
+        // It would've been nice to use bytemuck here, but it doesn't support const
+        // casting for enums.
+        unsafe { std::mem::transmute(value) }
+    }
+}
+
+// Used by bitfield-struct to convert into and out of the bitfield.
+impl StandardInstructionType {
+    const fn into_bits(self) -> u8 {
+        self as _
+    }
+
+    const fn from_bits(value: u8) -> Self {
+        match value {
+            0 => StandardInstructionType::Barrier,
+            1 => StandardInstructionType::Delay,
+            2 => StandardInstructionType::Measure,
+            3 => StandardInstructionType::Reset,
+            _ => panic!("unexpected standard instruction type!"),
+        }
+    }
+}
+
+impl PackedOperationType {
+    const fn from_bits(value: u8) -> Self {
+        match value {
+            0 => Self::StandardGate,
+            1 => Self::StandardInstruction,
+            2 => Self::PyGatePointer,
+            3 => Self::PyInstructionPointer,
+            4 => Self::PyOperationPointer,
+            _ => panic!("unexpected discriminant type!"),
+        }
+    }
+}
+
+/// The bitfield layout used for standard gates.
+#[bitfield(u64)]
+struct StandardGateBits {
+    #[bits(3, default = PackedOperationType::StandardGate, access = RO)]
+    discriminant: PackedOperationType,
+    #[bits(8)]
+    standard_gate: StandardGate,
+    #[bits(53)]
+    __: u64,
+}
+
+/// The bitfield layout used for standard instructions.
+#[bitfield(u64)]
+struct StandardInstructionBits {
+    #[bits(3, default = PackedOperationType::StandardInstruction, access = RO)]
+    discriminant: PackedOperationType,
+    #[bits(8)]
+    standard_instruction: StandardInstructionType,
+    #[bits(21)]
+    _pad1: u32,
+    #[bits(32)]
+    value: ImmediateValue,
+}
```

Comment:
It's odd to me that you went to all the trouble with the 32-bit pointers to make them aligned, even though they're the "terrible performance" path on platforms we don't even care much about, yet `StandardGate` and `StandardInstruction` (very much meant to be the absolute fastest paths) have their internal discriminant misaligned.

A second consideration: an alternative to having `StandardInstructionBits` forcibly define the bitlayout of `StandardInstruction` itself would be to have `StandardInstruction` have a `to_bits` method that is required to return a 61-bit value, and leave it up to `StandardInstruction`.  That would let the code be a lot more local to the rest of the type, while in this form, there's still co-operation necessary between `StandardInstruction` and `StandardInstructionBits`, but they're defined far away from each other.

Reply:
> It's odd to me that you went to all the trouble with the 32-bit pointers to make them aligned, even though they're the "terrible performance" path on platforms we don't even care much about, yet StandardGate and StandardInstruction (very much meant to be the absolute fastest paths) have their internal discriminant misaligned.

I did consider this, and we can certainly add a bit of padding to get them over the byte boundary. The 32-bit case was just what I was focusing on initially since I understood we would be impeding performance for that case the most. You didn't pad `StandardGate` originally, so I figured I wasn't making anything worse by deferring that.

Reply:
Yeah, I'm onboard with any potential modification to `StandardGate` being done in a separate PR - if nothing else, it'll let us see if the change is actually worth it.

When I wrote the initial PR, I think my thought process was "a bitshift will be so fast we don't notice it", and "we might need those padding bits later". Certainly worth challenging those assumptions - I didn't test them.